File: SudokuSolver_Java/src/main/sudokusolver/java/logic/tough/Swordfish.java
package sudokusolver.java.logic.tough;

import sudokusolver.java.Board;
import sudokusolver.java.Cell;
import sudokusolver.java.LocatedCandidate;
import sudokusolver.java.RemoveCandidates;
import sudokusolver.java.SudokuNumber;
import sudokusolver.java.Triple;
import sudokusolver.java.UnsolvedCell;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.IntFunction;
import java.util.function.ToIntFunction;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/*
 * https://www.sudokuwiki.org/Sword_Fish_Strategy
 *
 * For a triple of rows, if a candidate appears in two or three cells for each row and the candidate appears in exactly
 * three columns across the three rows, forming a three by three grid, then the candidate must be placed in three of the
 * nine cells. The candidate can be removed from cells which are in the three columns, but different rows.
 *
 * For a triple of columns, if a candidate appears in two or three cells for each column and the candidate appears in
 * exactly three rows across the three columns, forming a three by three grid, then the candidate must be placed in
 * three of the nine cells. The candidate can be removed from cells which are in the three rows, but different columns.
 */
public class Swordfish {
    public static List<RemoveCandidates> swordfish(Board<Cell> board) {
        return Arrays.stream(SudokuNumber.values())
                .flatMap(candidate -> {
                    var rowRemovals = swordfish(
                            candidate,
                            board.rows(),
                            board::getColumn,
                            Cell::column
                    );
                    var columnRemovals = swordfish(
                            candidate,
                            board.getColumns(),
                            board::getRow,
                            Cell::row
                    );
                    return Stream.concat(rowRemovals, columnRemovals);
                })
                .collect(LocatedCandidate.mergeToRemoveCandidates());
    }

    private static Stream<LocatedCandidate> swordfish(
            SudokuNumber candidate,
            List<List<Cell>> units,
            IntFunction<List<Cell>> getOtherUnit,
            ToIntFunction<Cell> getOtherUnitIndex
    ) {
        return units.stream()
                .collect(Triple.zipEveryTriple())
                .flatMap(triple -> {
                    var unitA = triple.first();
                    var unitB = triple.second();
                    var unitC = triple.third();
                    var aWithCandidate = unitA.stream()
                            .filter(UnsolvedCell.class::isInstance)
                            .map(UnsolvedCell.class::cast)
                            .filter(cell -> cell.candidates().contains(candidate))
                            .toList();
                    var bWithCandidate = unitB.stream()
                            .filter(UnsolvedCell.class::isInstance)
                            .map(UnsolvedCell.class::cast)
                            .filter(cell -> cell.candidates().contains(candidate))
                            .toList();
                    var cWithCandidate = unitC.stream()
                            .filter(UnsolvedCell.class::isInstance)
                            .map(UnsolvedCell.class::cast)
                            .filter(cell -> cell.candidates().contains(candidate))
                            .toList();
                    if ((aWithCandidate.size() == 2 || aWithCandidate.size() == 3) &&
                            (bWithCandidate.size() == 2 || bWithCandidate.size() == 3) &&
                            (cWithCandidate.size() == 2 || cWithCandidate.size() == 3)
                    ) {
                        var withCandidate = new ArrayList<>(aWithCandidate);
                        withCandidate.addAll(bWithCandidate);
                        withCandidate.addAll(cWithCandidate);
                        var otherUnitIndices = withCandidate.stream()
                                .map(getOtherUnitIndex::applyAsInt)
                                .collect(Collectors.toSet());
                        if (otherUnitIndices.size() == 3) {
                            return otherUnitIndices.stream()
                                    .flatMap(otherUnitIndex -> getOtherUnit.apply(otherUnitIndex).stream())
                                    .filter(UnsolvedCell.class::isInstance)
                                    .map(UnsolvedCell.class::cast)
                                    .filter(cell -> cell.candidates().contains(candidate) &&
                                            !withCandidate.contains(cell))
                                    .map(cell -> new LocatedCandidate(cell, candidate));
                        } else {
                            return Stream.empty();
                        }
                    } else {
                        return Stream.empty();
                    }
                });
    }
}

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/logic/tough/Swordfish.scala
package sudokusolver.scala.logic.tough

import sudokusolver.scala.*

/*
 * http://www.sudokuwiki.org/Sword_Fish_Strategy
 *
 * For a triple of rows, if a candidate appears in two or three cells for each row and the candidate appears in exactly
 * three columns across the three rows, forming a three by three grid, then the candidate must be placed in three of the
 * nine cells. The candidate can be removed from cells which are in the three columns, but different rows.
 *
 * For a triple of columns, if a candidate appears in two or three cells for each column and the candidate appears in
 * exactly three rows across the three columns, forming a three by three grid, then the candidate must be placed in
 * three of the nine cells. The candidate can be removed from cells which are in the three rows, but different columns.
 */
def swordfish(board: Board[Cell]): Seq[RemoveCandidates] =
  SudokuNumber.values.toSeq.flatMap { candidate =>

    def swordfish(units: Seq[Seq[Cell]], getOtherUnit: Int => Seq[Cell], getOtherUnitIndex: Cell => Int) =
      units.toIndexedSeq.zipEveryTriple.flatMap { (unitA, unitB, unitC) =>
        val aWithCandidate = unitA.collect { case cell: UnsolvedCell if cell.candidates.contains(candidate) => cell }
        val bWithCandidate = unitB.collect { case cell: UnsolvedCell if cell.candidates.contains(candidate) => cell }
        val cWithCandidate = unitC.collect { case cell: UnsolvedCell if cell.candidates.contains(candidate) => cell }
        if (2 to 3).contains(aWithCandidate.size) &&
          (2 to 3).contains(bWithCandidate.size) &&
          (2 to 3).contains(cWithCandidate.size) then
          val withCandidate = aWithCandidate ++ bWithCandidate ++ cWithCandidate
          val otherUnitIndices = withCandidate.map(getOtherUnitIndex(_)).toSet
          if otherUnitIndices.size == 3 then
            for
              cell <- otherUnitIndices.flatMap(getOtherUnit(_)).collect { case cell: UnsolvedCell => cell }
              if cell.candidates.contains(candidate) && !withCandidate.contains(cell)
            yield cell -> candidate
          else
            Nil
        else
          Nil
      }

    val rowRemovals = swordfish(board.rows, board.getColumn, _.column)
    val columnRemovals = swordfish(board.columns, board.getRow, _.row)
    rowRemovals ++ columnRemovals
  }.mergeToRemoveCandidates